tools/diff-processor/diff/diff.go

```diff
@@ -313,3 +326,42 @@ func setKey(set FieldSet) string {
 	slice := setToSortedSlice(set)
 	return strings.Join(slice, ",")
 }
+
+func (sd SchemaDiff) IsNewResource(resource string) bool {
+	var rcd ResourceConfigDiff
+	if rd, ok := (sd)[resource]; ok {
+		rcd = rd.ResourceConfig
+	}
+	if rcd.Old == nil && rcd.New != nil {
+		return true
+	}
+	return false
+}
+
+// IsFieldInNewNestedStructure determines if a field is part of a completely new nested structure
+func (sd SchemaDiff) IsFieldInNewNestedStructure(resourceName, fieldPath string) bool {
```

Comment:
Rather than checking whether the parent is new, we need to check whether there is at least one new ancestor (possibly a grandparent) that was optional. (I _think_ a new O+C with a nested required field is non-breaking; optional without computed definitely is. But if all the ancestors are required, then this will be a breaking change.)

Having a helper function for this makes a lot of sense since it's a complex operation that would be easy to get wrong - but I don't know that it makes sense as a method on the SchemaDiff object, since it should operate on a specific resource diff + field name.

Reply:
discussed offline